File: go/problems/67_add_binary.go
package main

import (
	"strconv"
)

func addBinary(a string, b string) string {
	carry := 0
	result := ""
	i, j := len(a)-1, len(b)-1

	for i >= 0 || j >= 0 {
		sum := carry

		if i >= 0 {
			sum += int(a[i] - '0')
			i--
		}

		if j >= 0 {
			sum += int(b[j] - '0')
			j--
		}

		carry = sum / 2
		char := sum % 2
		result = strconv.Itoa(char) + result
	}

	if carry > 0 {
		return "1" + result
	}
	return result
}

File: ts/problems/67-add-binary.ts
export function addBinary(a: string, b: string): string {
  let carry = 0;
  let result = "";
  let i = a.length - 1;
  let j = b.length - 1;

  while (i >= 0 || j >= 0) {
    let sum = carry;

    if (i >= 0) {
      sum += parseInt(a[i]);
      i--;
    }

    if (j >= 0) {
      sum += parseInt(b[j]);
      j--;
    }

    carry = Math.floor(sum / 2);
    let char = sum % 2;
    result = String(char) + result;
  }

  if (carry > 0) {
    return "1" + result;
  }
  return result;
}
